FWIO/RNTuple/plugins/RNTupleOutputFile.cc

```diff
@@ -0,0 +1,583 @@
+      //auto field = ROOT::Experimental::RFieldBase::Create("FileFormatVersion", "edm::FileFormatVersion").Unwrap();
+      //model->AddField(std::move(field));
+    }
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("FileIdentifier", "edm::FileID").Unwrap();
+      model->AddField(std::move(field));
+    }
+
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("IndexIntoFile", "edm::IndexIntoFile").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("MergeableRunProductMetadata",
+                                                          "edm::StoredMergeableRunProductMetadata")
+                       .Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field =
+          ROOT::Experimental::RFieldBase::Create("ProcessHistory", "std::vector<edm::ProcessHistory>").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("ProductRegistry", "edm::ProductRegistry").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field =
+          ROOT::Experimental::RFieldBase::Create("BranchIDLists", "std::vector<std::vector<unsigned int> >").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field =
+          ROOT::Experimental::RFieldBase::Create("ThinnedAssociationsHelper", "edm::ThinnedAssociationsHelper").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("ProductDependencies", "edm::BranchChildren").Unwrap();
+      model->AddField(std::move(field));
+    }
+
+    auto writeOptions = ROOT::Experimental::RNTupleWriteOptions();
+    writeOptions.SetCompression(convert(iConfig.compressionAlgo), iConfig.compressionLevel);
+    metaData_ = ROOT::Experimental::RNTupleWriter::Append(std::move(model), "MetaData", file_, writeOptions);
+  }
+
+  void RNTupleOutputFile::fillMetaData(BranchIDLists const& iBranchIDLists,
+                                       ThinnedAssociationsHelper const& iThinnedHelper,
+                                       ProductRegistry const& iReg) {
+    auto rentry = metaData_->CreateEntry();
+
+    FileID id(createGlobalIdentifier());
+
+    rentry->BindRawPtr("FileIdentifier", &id);
+
+    indexIntoFile_.sortVector_Run_Or_Lumi_Entries();
+    rentry->BindRawPtr("IndexIntoFile", &indexIntoFile_);
+
+    ProcessHistoryVector procHistoryVector;
+    for (auto const& ph : processHistoryRegistry_) {
+      procHistoryVector.push_back(ph.second);
+    }
+    rentry->BindRawPtr("ProcessHistory", &procHistoryVector);
+
+    // Make a local copy of the ProductRegistry, removing any transient or pruned products.
+    using ProductList = ProductRegistry::ProductList;
+    ProductRegistry pReg(iReg.productList());
+    ProductList& pList = const_cast<ProductList&>(pReg.productList());
```

Comment:
Should we think of adding a non-`const` overload to `ProductRegistry`?

Reply:
Possibly. I think the real 'problem' is RNTuple requires passing a `void*` to `BindRawPtr` even if the object is only being read.